src/pseudopeople/noise_functions.py

```diff
@@ -608,37 +585,140 @@ def make_ocr_errors(
+    )
+
+    rng = np.random.default_rng(seed=get_hash(f"{randomness_stream.seed}_{addl_key}"))
+
+    # NOTE: Somewhat surprisingly, this seemed to perform better using Python string types than NumPy types.
+    # Perhaps worth more investigation in the future.
+    # Note that each item in result can be up to the number of characters of the *longest* corrupted token.
+    result = np.empty(same_len_col_exploded.shape, dtype=str).astype("object")
+    # Tracks the next character index where we have to pay attention to a given string.
+    # When we corrupt a token, we never corrupt again within that same source token, so we
+    # don't have to consider that string again until we get to the end of it.
+    next_due = np.zeros(len(column))
+    # We proceed by character through all the strings at once
+    for i in range(same_len_col_exploded.shape[1]):
+        error_introduced = np.zeros(len(column), dtype=bool)
+        # Longer tokens to-be-corrupted take precedence over shorter ones
+        for token_length in range(max_token_length, 0, -1):
+            if i + token_length > same_len_col_exploded.shape[1]:
+                continue
+
+            due = ~error_introduced & (next_due <= i)
+
+            # Is the string already over at this index?
+            # If so, we can ignore it.
+            long_enough = np.zeros(len(column), dtype=bool)
+            long_enough[due] = i + token_length <= lengths[due]
+            if long_enough.sum() == 0:
+                continue
+
+            # Collect the tokens of the given length that start at this index.
+            tokens = np.empty(len(column), dtype=f"U{token_length}")
+            tokens[long_enough] = (
+                same_len_col_exploded[long_enough, i : (i + token_length)]
+                .view(f"U{token_length}")
+                .reshape(long_enough.sum())
+            )
+
+            # Are these tokens that have corruptions?
+            eligible = np.zeros(len(column), dtype=bool)
+            eligible[long_enough] = (
+                pd.Series(tokens[long_enough])
+                .isin(errors_eligible_tokens[token_length])
+                .values
+            )
+            if eligible.sum() == 0:
+                continue
+
+            # If so, should we corrupt them?
+            corrupted = np.zeros(len(column), dtype=bool)
+            corrupted[eligible] = rng.random(eligible.sum()) < token_probability
+            if corrupted.sum() == 0:
+                continue
+
+            # If so, which corruption should we choose?
+            to_corrupt = tokens[corrupted]
+            corrupted_token_index = np.floor(
+                rng.random(corrupted.sum()) * number_of_options.loc[to_corrupt].to_numpy()
+            )
+            # Get the actual string corresponding to the corrupted token chosen.
+            # Numpy does not have an efficient way to do this merge operation
+            result[corrupted, i] = (
+                pd.DataFrame(
```

Comment:
Consider creating a pandas series to "flip" the index of the np array (i.e. array("a", "b", "c") goes to a Series with index a, b, c and values, 1, 2, 3.